jobserver/views/projects.py

```diff
@@ -153,11 +154,30 @@ def dispatch(self, request, *args, **kwargs):
     def form_valid(self, form):
         users = form.cleaned_data["users"]
 
+        failed_to_invite = []
+
         for user in users:
-            ProjectInvitation.objects.create(
-                created_by=self.request.user,
-                project=self.project,
-                user=user,
+            try:
+                with transaction.atomic():
+                    invite = ProjectInvitation.objects.create(
+                        created_by=self.request.user,
+                        project=self.project,
+                        user=user,
+                    )
+                    send_project_invite_email(
+                        user.notifications_email, self.project, invite
+                    )
+            except Exception:
+                failed_to_invite.append(user)
+
+        if failed_to_invite:
+            # tell the user about each failed invite in a single message
+            users = [f"<li>{u.username}</li>" for u in failed_to_invite]
+            users = f"<ul>{''.join(users)}</ul>"
+
+            messages.error(
+                self.request,
+                f"<p>Failed to invite {len(failed_to_invite)} User(s):</p>{users}<p>Please try again.</p>",
```

Comment:
Is this vulnerable to XSS?

Reply:
I don't _think_ so…

`users` coming from `form.cleaned_data` is a QuerySet of User instances so even if someone managed to push modified User IDs from the form the backend, it's still only using data from the database which GitHub gave us.

Unless I'm missing something else?

Reply:
We're (a) trusting that GitHub doesn't allow usernames with funny characters in and (b) assuming that we never get usernames from anywhere else.

(a) is true now, and is likely to be true in future, but given how easy this is to protect against, I think we should do so now.

Reply:
Fair point!